File: Sources/PyCal/Calculator/ExpressionEvaluator.swift
import Foundation

struct EvaluationResult: Equatable {
    let value: Double
    let assignment: String?
}

enum ExpressionError: LocalizedError, Equatable {
    case emptyExpression
    case invalidCharacter(Character, Int)
    case invalidNumber(String)
    case unexpectedToken(String)
    case missingClosingParenthesis
    case unknownVariable(String)
    case unknownFunction(String)
    case wrongArgumentCount(name: String, expected: String)
    case divisionByZero
    case invalidAssignment
    case nonFiniteResult

    var errorDescription: String? {
        switch self {
        case .emptyExpression:
            "请输入计算公式"
        case let .invalidCharacter(character, position):
            "第 \(position + 1) 个字符“\(character)”无法识别"
        case let .invalidNumber(number):
            "“\(number)”不是有效数字"
        case let .unexpectedToken(token):
            "“\(token)”出现在了不正确的位置"
        case .missingClosingParenthesis:
            "缺少右括号 )"
        case let .unknownVariable(name):
            "变量“\(name)”尚未定义"
        case let .unknownFunction(name):
            "函数“\(name)”暂不支持"
        case let .wrongArgumentCount(name, expected):
            "函数 \(name) 需要\(expected)个参数"
        case .divisionByZero:
            "不能除以 0"
        case .invalidAssignment:
            "赋值左侧必须是变量名，例如 tax = 0.06"
        case .nonFiniteResult:
            "计算结果超出可表示范围"
        }
    }
}

/// A small, deterministic Python-style numeric expression evaluator.
///
/// Supported operators: `+ - * / // % **`, parentheses, assignments and a
/// focused set of math functions. `^` is accepted as a calculator-friendly
/// alias for exponentiation.
struct ExpressionEvaluator {
    private let variables: [String: Double]

    init(variables: [String: Double] = [:]) {
        self.variables = variables
    }

    func evaluate(_ source: String) throws -> EvaluationResult {
        let normalized = source
            .replacingOccurrences(of: "×", with: "*")
            .replacingOccurrences(of: "÷", with: "/")
            .replacingOccurrences(of: "（", with: "(")
            .replacingOccurrences(of: "）", with: ")")
            .replacingOccurrences(of: "，", with: ",")

        var lexer = Lexer(normalized)
        let tokens = try lexer.tokenize()
        var parser = Parser(tokens: tokens, variables: variables)
        return try parser.parse()
    }
}

private enum Token: Equatable, CustomStringConvertible {
    case number(Double)
    case identifier(String)
    case plus
    case minus
    case multiply
    case divide
    case floorDivide
    case modulo
    case power
    case leftParen
    case rightParen
    case comma
    case assign
    case end

    var description: String {
        switch self {
        case let .number(value): String(value)
        case let .identifier(name): name
        case .plus: "+"
        case .minus: "-"
        case .multiply: "*"
        case .divide: "/"
        case .floorDivide: "//"
        case .modulo: "%"
        case .power: "**"
        case .leftParen: "("
        case .rightParen: ")"
        case .comma: ","
        case .assign: "="
        case .end: "公式末尾"
        }
    }
}

private struct Lexer {
    let characters: [Character]
    var index = 0

    init(_ source: String) {
        characters = Array(source)
    }

    mutating func tokenize() throws -> [Token] {
        var tokens: [Token] = []

        while index < characters.count {
            let character = characters[index]

            if character.isWhitespace {
                index += 1
                continue
            }

            if character.isNumber || character == "." {
                tokens.append(try readNumber())
                continue
            }

            if character.isLetter || character == "_" {
                tokens.append(readIdentifier())
                continue
            }

            switch character {
            case "+":
                tokens.append(.plus)
            case "-":
                tokens.append(.minus)
            case "*":
                if peekNext() == "*" {
                    index += 1
                    tokens.append(.power)
                } else {
                    tokens.append(.multiply)
                }
            case "/":
                if peekNext() == "/" {
                    index += 1
                    tokens.append(.floorDivide)
                } else {
                    tokens.append(.divide)
                }
            case "%":
                tokens.append(.modulo)
            case "^":
                tokens.append(.power)
            case "(":
                tokens.append(.leftParen)
            case ")":
                tokens.append(.rightParen)
            case ",":
                tokens.append(.comma)
            case "=":
                tokens.append(.assign)
            default:
                throw ExpressionError.invalidCharacter(character, index)
            }
            index += 1
        }

        tokens.append(.end)
        return tokens
    }

    private mutating func readNumber() throws -> Token {
        let start = index
        var sawDot = false
        var sawExponent = false

        while index < characters.count {
            let current = characters[index]
            if current.isNumber || current == "_" {
                index += 1
            } else if current == ".", !sawDot, !sawExponent {
                sawDot = true
                index += 1
            } else if (current == "e" || current == "E"), !sawExponent {
                sawExponent = true
                index += 1
                if index < characters.count, characters[index] == "+" || characters[index] == "-" {
                    index += 1
                }
            } else {
                break
            }
        }

        let raw = String(characters[start..<index])
        let cleaned = raw.replacingOccurrences(of: "_", with: "")
        guard cleaned != ".", let value = Double(cleaned) else {
            throw ExpressionError.invalidNumber(raw)
        }
        return .number(value)
    }

    private mutating func readIdentifier() -> Token {
        let start = index
        while index < characters.count {
            let current = characters[index]
            guard current.isLetter || current.isNumber || current == "_" else { break }
            index += 1
        }
        return .identifier(String(characters[start..<index]))
    }

    private func peekNext() -> Character? {
        let next = index + 1
        return next < characters.count ? characters[next] : nil
    }
}

private struct Parser {
    let tokens: [Token]
    let variables: [String: Double]
    var index = 0

    mutating func parse() throws -> EvaluationResult {
        guard current != .end else { throw ExpressionError.emptyExpression }

        var assignment: String?
        if case let .identifier(name) = current, peek == .assign {
            guard Self.isValidVariableName(name) else { throw ExpressionError.invalidAssignment }
            assignment = name
            advance()
            advance()
        } else if tokens.dropLast().contains(.assign) {
            throw ExpressionError.invalidAssignment
        }

        let value = try parseAdditive()
        guard current == .end else {
            if current == .rightParen { throw ExpressionError.unexpectedToken(")") }
            throw ExpressionError.unexpectedToken(current.description)
        }
        guard value.isFinite else { throw ExpressionError.nonFiniteResult }
        return EvaluationResult(value: value, assignment: assignment)
    }

    private mutating func parseAdditive() throws -> Double {
        var value = try parseMultiplicative()
        while true {
            if match(.plus) {
                value += try parseMultiplicative()
            } else if match(.minus) {
                value -= try parseMultiplicative()
            } else {
                return value
            }
        }
    }

    private mutating func parseMultiplicative() throws -> Double {
        var value = try parseUnary()
        while true {
            if match(.multiply) {
                value *= try parseUnary()
            } else if match(.divide) {
                let divisor = try parseUnary()
                guard divisor != 0 else { throw ExpressionError.divisionByZero }
                value /= divisor
            } else if match(.floorDivide) {
                let divisor = try parseUnary()
                guard divisor != 0 else { throw ExpressionError.divisionByZero }
                value = floor(value / divisor)
            } else if match(.modulo) {
                let divisor = try parseUnary()
                guard divisor != 0 else { throw ExpressionError.divisionByZero }
                value = value - floor(value / divisor) * divisor
            } else {
                return value
            }
        }
    }

    // Python gives exponentiation higher precedence than unary minus:
    // -2**2 == -(2**2), while 2**-2 is valid.
    private mutating func parseUnary() throws -> Double {
        if match(.plus) { return try parseUnary() }
        if match(.minus) { return -(try parseUnary()) }
        return try parsePower()
    }

    private mutating func parsePower() throws -> Double {
        let base = try parsePrimary()
        if match(.power) {
            return pow(base, try parseUnary())
        }
        return base
    }

    private mutating func parsePrimary() throws -> Double {
        switch current {
        case let .number(value):
            advance()
            return value
        case let .identifier(name):
            advance()
            if match(.leftParen) {
                return try parseFunction(name)
            }
            if let value = variables[name] { return value }
            switch name.lowercased() {
            case "pi": return .pi
            case "e": return M_E
            case "tau": return 2 * .pi
            default: throw ExpressionError.unknownVariable(name)
            }
        case .leftParen:
            advance()
            let value = try parseAdditive()
            guard match(.rightParen) else { throw ExpressionError.missingClosingParenthesis }
            return value
        default:
            throw ExpressionError.unexpectedToken(current.description)
        }
    }

    private mutating func parseFunction(_ name: String) throws -> Double {
        var arguments: [Double] = []
        if current != .rightParen {
            repeat {
                arguments.append(try parseAdditive())
            } while match(.comma)
        }
        guard match(.rightParen) else { throw ExpressionError.missingClosingParenthesis }

        let lowered = name.lowercased()
        switch lowered {
        case "abs": return try unaryFunction(name, arguments, abs)
        case "sqrt": return try unaryFunction(name, arguments, sqrt)
        case "sin": return try unaryFunction(name, arguments, sin)
        case "cos": return try unaryFunction(name, arguments, cos)
        case "tan": return try unaryFunction(name, arguments, tan)
        case "asin": return try unaryFunction(name, arguments, asin)
        case "acos": return try unaryFunction(name, arguments, acos)
        case "atan": return try unaryFunction(name, arguments, atan)
        case "exp": return try unaryFunction(name, arguments, exp)
        case "ln": return try unaryFunction(name, arguments, log)
        case "log10": return try unaryFunction(name, arguments, log10)
        case "floor": return try unaryFunction(name, arguments, floor)
        case "ceil": return try unaryFunction(name, arguments, ceil)
        case "round":
            guard arguments.count == 1 || arguments.count == 2 else {
                throw ExpressionError.wrongArgumentCount(name: name, expected: " 1 或 2 ")
            }
            if arguments.count == 1 { return arguments[0].rounded(.toNearestOrEven) }
            let places = arguments[1]
            guard places.isFinite, places.rounded() == places, abs(places) <= 308 else {
                throw ExpressionError.wrongArgumentCount(name: name, expected: "第二个参数需为整数位数")
            }
            let factor = pow(10, places)
            return (arguments[0] * factor).rounded(.toNearestOrEven) / factor
        case "pow":
            guard arguments.count == 2 else {
                throw ExpressionError.wrongArgumentCount(name: name, expected: " 2 ")
            }
            return pow(arguments[0], arguments[1])
        case "log":
            guard arguments.count == 1 || arguments.count == 2 else {
                throw ExpressionError.wrongArgumentCount(name: name, expected: " 1 或 2 ")
            }
            return arguments.count == 1 ? log(arguments[0]) : log(arguments[0]) / log(arguments[1])
        case "min":
            guard !arguments.isEmpty else {
                throw ExpressionError.wrongArgumentCount(name: name, expected: "至少 1 ")
            }
            return arguments.min()!
        case "max":
            guard !arguments.isEmpty else {
                throw ExpressionError.wrongArgumentCount(name: name, expected: "至少 1 ")
            }
            return arguments.max()!
        default:
            throw ExpressionError.unknownFunction(name)
        }
    }

    private func unaryFunction(
        _ name: String,
        _ arguments: [Double],
        _ operation: (Double) -> Double
    ) throws -> Double {
        guard arguments.count == 1 else {
            throw ExpressionError.wrongArgumentCount(name: name, expected: " 1 ")
        }
        return operation(arguments[0])
    }

    private static func isValidVariableName(_ name: String) -> Bool {
        guard let first = name.first, first.isLetter || first == "_" else { return false }
        return name.allSatisfy { $0.isLetter || $0.isNumber || $0 == "_" }
    }

    private var current: Token { tokens[index] }
    private var peek: Token { tokens[min(index + 1, tokens.count - 1)] }

    private mutating func advance() {
        if index < tokens.count - 1 { index += 1 }
    }

    private mutating func match(_ token: Token) -> Bool {
        guard current == token else { return false }
        advance()
        return true
    }
}

File: lib/calculator/expression_evaluator.dart
import 'dart:math' as math;

/// Result of evaluating a numeric expression, with an optional assignment target.
class EvaluationResult {
  const EvaluationResult({required this.value, this.assignment});

  final double value;
  final String? assignment;

  @override
  bool operator ==(Object other) =>
      other is EvaluationResult &&
      other.value == value &&
      other.assignment == assignment;

  @override
  int get hashCode => Object.hash(value, assignment);
}

/// Deterministic errors that match the Swift evaluator's messages.
class ExpressionError implements Exception {
  const ExpressionError._(this.kind, this.message, [this.details]);

  final String kind;
  final String message;
  final Object? details;

  factory ExpressionError.emptyExpression() =>
      const ExpressionError._('emptyExpression', '请输入计算公式');

  factory ExpressionError.invalidCharacter(String character, int position) =>
      ExpressionError._(
        'invalidCharacter',
        '第 ${position + 1} 个字符“$character”无法识别',
        (character, position),
      );

  factory ExpressionError.invalidNumber(String number) =>
      ExpressionError._('invalidNumber', '“$number”不是有效数字', number);

  factory ExpressionError.unexpectedToken(String token) =>
      ExpressionError._('unexpectedToken', '“$token”出现在了不正确的位置', token);

  factory ExpressionError.missingClosingParenthesis() =>
      const ExpressionError._('missingClosingParenthesis', '缺少右括号 )');

  factory ExpressionError.unknownVariable(String name) =>
      ExpressionError._('unknownVariable', '变量“$name”尚未定义', name);

  factory ExpressionError.unknownFunction(String name) =>
      ExpressionError._('unknownFunction', '函数“$name”暂不支持', name);

  factory ExpressionError.wrongArgumentCount(String name, String expected) =>
      ExpressionError._(
        'wrongArgumentCount',
        '函数 $name 需要$expected个参数',
        (name, expected),
      );

  factory ExpressionError.divisionByZero() =>
      const ExpressionError._('divisionByZero', '不能除以 0');

  factory ExpressionError.invalidAssignment() =>
      const ExpressionError._('invalidAssignment', '赋值左侧必须是变量名，例如 tax = 0.06');

  factory ExpressionError.nonFiniteResult() =>
      const ExpressionError._('nonFiniteResult', '计算结果超出可表示范围');

  @override
  String toString() => message;

  @override
  bool operator ==(Object other) =>
      other is ExpressionError && other.kind == kind && other.details == details;

  @override
  int get hashCode => Object.hash(kind, details);
}

bool isValidVariableName(String name) {
  if (name.isEmpty) return false;
  final chars = _chars(name);
  if (!_isIdentStart(chars.first)) return false;
  return chars.every(_isIdentContinue);
}

/// A small, deterministic Python-style numeric expression evaluator.
///
/// Supported operators: `+ - * / // % **`, parentheses, assignments and a
/// focused set of math functions. `^` is accepted as a calculator-friendly
/// alias for exponentiation.
class ExpressionEvaluator {
  const ExpressionEvaluator({this.variables = const {}});

  final Map<String, double> variables;

  EvaluationResult evaluate(String source) {
    final normalized = source
        .replaceAll('×', '*')
        .replaceAll('÷', '/')
        .replaceAll('（', '(')
        .replaceAll('）', ')')
        .replaceAll('，', ',');

    final tokens = _Lexer(normalized).tokenize();
    return _Parser(tokens: tokens, variables: variables).parse();
  }
}

enum _TokenKind {
  number,
  identifier,
  plus,
  minus,
  multiply,
  divide,
  floorDivide,
  modulo,
  power,
  leftParen,
  rightParen,
  comma,
  assign,
  end,
}

class _Token {
  const _Token(this.kind, {this.number, this.identifier});

  final _TokenKind kind;
  final double? number;
  final String? identifier;

  String get description {
    switch (kind) {
      case _TokenKind.number:
        return '$number';
      case _TokenKind.identifier:
        return identifier ?? '';
      case _TokenKind.plus:
        return '+';
      case _TokenKind.minus:
        return '-';
      case _TokenKind.multiply:
        return '*';
      case _TokenKind.divide:
        return '/';
      case _TokenKind.floorDivide:
        return '//';
      case _TokenKind.modulo:
        return '%';
      case _TokenKind.power:
        return '**';
      case _TokenKind.leftParen:
        return '(';
      case _TokenKind.rightParen:
        return ')';
      case _TokenKind.comma:
        return ',';
      case _TokenKind.assign:
        return '=';
      case _TokenKind.end:
        return '公式末尾';
    }
  }

  @override
  bool operator ==(Object other) =>
      other is _Token &&
      other.kind == kind &&
      other.number == number &&
      other.identifier == identifier;

  @override
  int get hashCode => Object.hash(kind, number, identifier);
}

class _Lexer {
  _Lexer(String source) : characters = _chars(source);

  final List<String> characters;
  var index = 0;

  List<_Token> tokenize() {
    final tokens = <_Token>[];

    while (index < characters.length) {
      final character = characters[index];

      if (_isWhitespace(character)) {
        index += 1;
        continue;
      }

      if (_isDigit(character) || character == '.') {
        tokens.add(_readNumber());
        continue;
      }

      if (_isIdentStart(character)) {
        tokens.add(_readIdentifier());
        continue;
      }

      switch (character) {
        case '+':
          tokens.add(const _Token(_TokenKind.plus));
        case '-':
          tokens.add(const _Token(_TokenKind.minus));
        case '*':
          if (_peekNext() == '*') {
            index += 1;
            tokens.add(const _Token(_TokenKind.power));
          } else {
            tokens.add(const _Token(_TokenKind.multiply));
          }
        case '/':
          if (_peekNext() == '/') {
            index += 1;
            tokens.add(const _Token(_TokenKind.floorDivide));
          } else {
            tokens.add(const _Token(_TokenKind.divide));
          }
        case '%':
          tokens.add(const _Token(_TokenKind.modulo));
        case '^':
          tokens.add(const _Token(_TokenKind.power));
        case '(':
          tokens.add(const _Token(_TokenKind.leftParen));
        case ')':
          tokens.add(const _Token(_TokenKind.rightParen));
        case ',':
          tokens.add(const _Token(_TokenKind.comma));
        case '=':
          tokens.add(const _Token(_TokenKind.assign));
        default:
          throw ExpressionError.invalidCharacter(character, index);
      }
      index += 1;
    }

    tokens.add(const _Token(_TokenKind.end));
    return tokens;
  }

  _Token _readNumber() {
    final start = index;
    var sawDot = false;
    var sawExponent = false;

    while (index < characters.length) {
      final current = characters[index];
      if (_isDigit(current) || current == '_') {
        index += 1;
      } else if (current == '.' && !sawDot && !sawExponent) {
        sawDot = true;
        index += 1;
      } else if ((current == 'e' || current == 'E') && !sawExponent) {
        sawExponent = true;
        index += 1;
        if (index < characters.length &&
            (characters[index] == '+' || characters[index] == '-')) {
          index += 1;
        }
      } else {
        break;
      }
    }

    final raw = characters.sublist(start, index).join();
    final cleaned = raw.replaceAll('_', '');
    final value = double.tryParse(cleaned);
    if (cleaned == '.' || value == null) {
      throw ExpressionError.invalidNumber(raw);
    }
    return _Token(_TokenKind.number, number: value);
  }

  _Token _readIdentifier() {
    final start = index;
    while (index < characters.length) {
      final current = characters[index];
      if (!_isIdentContinue(current)) break;
      index += 1;
    }
    return _Token(
      _TokenKind.identifier,
      identifier: characters.sublist(start, index).join(),
    );
  }

  String? _peekNext() {
    final next = index + 1;
    return next < characters.length ? characters[next] : null;
  }
}

class _Parser {
  _Parser({required this.tokens, required this.variables});

  final List<_Token> tokens;
  final Map<String, double> variables;
  var index = 0;

  EvaluationResult parse() {
    if (current.kind == _TokenKind.end) {
      throw ExpressionError.emptyExpression();
    }

    String? assignment;
    if (current.kind == _TokenKind.identifier && peek.kind == _TokenKind.assign) {
      final name = current.identifier!;
      if (!isValidVariableName(name)) {
        throw ExpressionError.invalidAssignment();
      }
      assignment = name;
      advance();
      advance();
    } else if (tokens
        .take(tokens.length - 1)
        .any((token) => token.kind == _TokenKind.assign)) {
      throw ExpressionError.invalidAssignment();
    }

    final value = parseAdditive();
    if (current.kind != _TokenKind.end) {
      if (current.kind == _TokenKind.rightParen) {
        throw ExpressionError.unexpectedToken(')');
      }
      throw ExpressionError.unexpectedToken(current.description);
    }
    if (!value.isFinite) {
      throw ExpressionError.nonFiniteResult();
    }
    return EvaluationResult(value: value, assignment: assignment);
  }

  double parseAdditive() {
    var value = parseMultiplicative();
    while (true) {
      if (match(_TokenKind.plus)) {
        value += parseMultiplicative();
      } else if (match(_TokenKind.minus)) {
        value -= parseMultiplicative();
      } else {
        return value;
      }
    }
  }

  double parseMultiplicative() {
    var value = parseUnary();
    while (true) {
      if (match(_TokenKind.multiply)) {
        value *= parseUnary();
      } else if (match(_TokenKind.divide)) {
        final divisor = parseUnary();
        if (divisor == 0) throw ExpressionError.divisionByZero();
        value /= divisor;
      } else if (match(_TokenKind.floorDivide)) {
        final divisor = parseUnary();
        if (divisor == 0) throw ExpressionError.divisionByZero();
        value = (value / divisor).floorToDouble();
      } else if (match(_TokenKind.modulo)) {
        final divisor = parseUnary();
        if (divisor == 0) throw ExpressionError.divisionByZero();
        value = value - (value / divisor).floorToDouble() * divisor;
      } else {
        return value;
      }
    }
  }

  // Python gives exponentiation higher precedence than unary minus:
  // -2**2 == -(2**2), while 2**-2 is valid.
  double parseUnary() {
    if (match(_TokenKind.plus)) return parseUnary();
    if (match(_TokenKind.minus)) return -parseUnary();
    return parsePower();
  }

  double parsePower() {
    final base = parsePrimary();
    if (match(_TokenKind.power)) {
      return math.pow(base, parseUnary()).toDouble();
    }
    return base;
  }

  double parsePrimary() {
    switch (current.kind) {
      case _TokenKind.number:
        final value = current.number!;
        advance();
        return value;
      case _TokenKind.identifier:
        final name = current.identifier!;
        advance();
        if (match(_TokenKind.leftParen)) {
          return parseFunction(name);
        }
        final variable = variables[name];
        if (variable != null) return variable;
        switch (name.toLowerCase()) {
          case 'pi':
            return math.pi;
          case 'e':
            return math.e;
          case 'tau':
            return 2 * math.pi;
          default:
            throw ExpressionError.unknownVariable(name);
        }
      case _TokenKind.leftParen:
        advance();
        final value = parseAdditive();
        if (!match(_TokenKind.rightParen)) {
          throw ExpressionError.missingClosingParenthesis();
        }
        return value;
      default:
        throw ExpressionError.unexpectedToken(current.description);
    }
  }

  double parseFunction(String name) {
    final arguments = <double>[];
    if (current.kind != _TokenKind.rightParen) {
      do {
        arguments.add(parseAdditive());
      } while (match(_TokenKind.comma));
    }
    if (!match(_TokenKind.rightParen)) {
      throw ExpressionError.missingClosingParenthesis();
    }

    final lowered = name.toLowerCase();
    switch (lowered) {
      case 'abs':
        return _unaryFunction(name, arguments, (x) => x.abs());
      case 'sqrt':
        return _unaryFunction(name, arguments, math.sqrt);
      case 'sin':
        return _unaryFunction(name, arguments, math.sin);
      case 'cos':
        return _unaryFunction(name, arguments, math.cos);
      case 'tan':
        return _unaryFunction(name, arguments, math.tan);
      case 'asin':
        return _unaryFunction(name, arguments, math.asin);
      case 'acos':
        return _unaryFunction(name, arguments, math.acos);
      case 'atan':
        return _unaryFunction(name, arguments, math.atan);
      case 'exp':
        return _unaryFunction(name, arguments, math.exp);
      case 'ln':
        return _unaryFunction(name, arguments, math.log);
      case 'log10':
        return _unaryFunction(name, arguments, (x) => math.log(x) / math.ln10);
      case 'floor':
        return _unaryFunction(name, arguments, (x) => x.floorToDouble());
      case 'ceil':
        return _unaryFunction(name, arguments, (x) => x.ceilToDouble());
      case 'round':
        if (arguments.length != 1 && arguments.length != 2) {
          throw ExpressionError.wrongArgumentCount(name, ' 1 或 2 ');
        }
        if (arguments.length == 1) {
          return _roundToNearestEven(arguments[0]);
        }
        final places = arguments[1];
        if (!places.isFinite ||
            places.roundToDouble() != places ||
            places.abs() > 308) {
          throw ExpressionError.wrongArgumentCount(name, '第二个参数需为整数位数');
        }
        final factor = math.pow(10, places).toDouble();
        return _roundToNearestEven(arguments[0] * factor) / factor;
      case 'pow':
        if (arguments.length != 2) {
          throw ExpressionError.wrongArgumentCount(name, ' 2 ');
        }
        return math.pow(arguments[0], arguments[1]).toDouble();
      case 'log':
        if (arguments.length != 1 && arguments.length != 2) {
          throw ExpressionError.wrongArgumentCount(name, ' 1 或 2 ');
        }
        return arguments.length == 1
            ? math.log(arguments[0])
            : math.log(arguments[0]) / math.log(arguments[1]);
      case 'min':
        if (arguments.isEmpty) {
          throw ExpressionError.wrongArgumentCount(name, '至少 1 ');
        }
        return arguments.reduce(math.min);
      case 'max':
        if (arguments.isEmpty) {
          throw ExpressionError.wrongArgumentCount(name, '至少 1 ');
        }
        return arguments.reduce(math.max);
      default:
        throw ExpressionError.unknownFunction(name);
    }
  }

  double _unaryFunction(
    String name,
    List<double> arguments,
    double Function(double) operation,
  ) {
    if (arguments.length != 1) {
      throw ExpressionError.wrongArgumentCount(name, ' 1 ');
    }
    return operation(arguments[0]);
  }

  _Token get current => tokens[index];
  _Token get peek => tokens[math.min(index + 1, tokens.length - 1)];

  void advance() {
    if (index < tokens.length - 1) index += 1;
  }

  bool match(_TokenKind kind) {
    if (current.kind != kind) return false;
    advance();
    return true;
  }
}

double _roundToNearestEven(double value) {
  if (!value.isFinite) return value;
  final sign = value.isNegative ? -1.0 : 1.0;
  final magnitude = value.abs();
  final integer = magnitude.floorToDouble();
  final fraction = magnitude - integer;
  double rounded;
  if (fraction > 0.5) {
    rounded = integer + 1;
  } else if (fraction < 0.5) {
    rounded = integer;
  } else {
    rounded = integer % 2 == 0 ? integer : integer + 1;
  }
  return rounded * sign;
}

List<String> _chars(String source) =>
    source.runes.map(String.fromCharCode).toList(growable: false);

bool _isWhitespace(String character) => RegExp(r'^\s$').hasMatch(character);

bool _isDigit(String character) => RegExp(r'^\p{Nd}$', unicode: true).hasMatch(character);

bool _isIdentStart(String character) =>
    character == '_' || RegExp(r'^\p{L}$', unicode: true).hasMatch(character);

bool _isIdentContinue(String character) =>
    _isIdentStart(character) || _isDigit(character);
